### PR TITLE
BFD-1671: Add checkstyle and javadoc style guide

### PR DESCRIPTION
**Pull request by lsmitchell**
**JIRA Ticket:**
[BFD-1671](https://jira.cms.gov/browse/BFD-1671)

**User Story or Bug Summary:**

As a BFD user I want the documentation of the java code to be consistent and clear so that I can more easily make changes and maintain the codebase.

---

### What Does This PR Do?

- Added checkstyle to the build by adding a maven plugin, which will check enabled projects within BFD for proper javadoc styling (and that docs exist where they should)
- Fixed bfd-shared-utils to be in compliance with the new javadoc rules
- Added checkstyle.xml to the root of the apps folder which will be used to check all checkstyle-enabled projects
- Disabled checkstyle for projects other than shared-utils so checkstyle is not run on them until compliance is achieved (tech debt)
- Added a style guide to the BFD documentation which reviews the javadoc rules to be enforced (most of which are also enforced by the build using checkstyle)

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.
* Verify style guide matches what was agreed upon by the team
* Verify checkstyle.xml is understandable

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: 
New dependency for checkstyle via the maven checkstyle plugin:
https://mvnrepository.com/artifact/org.apache.maven.plugins/maven-checkstyle-plugin
https://maven.apache.org/plugins/maven-checkstyle-plugin/

    * Does this PR add any new software dependencies? 
      * [X] Yes
      * [ ] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [X] Yes
      * [ ] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by StewGoin — APPROVED** [review]
No concerns from my end, just make sure to update any software inventory list that may be applicable (not certain if we have one that details down to maven plugins tbh).

File: apps/checkstyle.xml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        <!-- Ensures non-local variables have docs.
+             Note that ignoreNamePattern can be used to ignore patterns of names
+             via RegEx if we want to avoid documenting certain variable names.
+        -->
+        <module name="JavadocVariable"/>
+
+        <!-- Style Rules -->
+        <!-- Good Practice rule for avoiding importing more than is required from a library -->
+        <module name="AvoidStarImport">
+            <property name="severity" value="warning"/>
```
Comment (christrotteradhoc) [inline apps/checkstyle.xml:37]: while I agree with this practice, we should discuss it in the sync in the next round of style changes before adding it.

Comment (lsmitchell) [reply apps/checkstyle.xml]: I left it as warn so it wouldnt stop a build, but I'll just remove it for now so there's no confusion.

**Review by christrotteradhoc — APPROVED** [review]

File: docs/style/style-guide.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+- All methods and constructors of all scopes must have a javadoc
+  - Must have a description that explains what the method does, and any useful contextual/historical information that may help maintainability
+  - All method parameters should be documented with the ```@param``` tag, and a description of what the parameter is
+    - If ```null``` is explicitly disallowed for a parameter, it should be documented
+  - Method returns should be documented with the ```@returns``` tag and a description of what is returned (and how if multiple returns under various conditions)
+    - If the method can return ```null``` it should be noted and under what condition this may happen
+  - Method overrides or interface implementations should have a javadoc that is simply ```{@inheritDoc}``` unless the overridden method's functionality diverges enough to need additional documentation
+    - This includes ```toString```, ```hashCode```, etc.
+- All classes and interfaces of all scopes must have a javadoc 
+- All class-level fields of all scopes must have a javadoc
```
Comment (mackec1) [inline docs/style/style-guide.md:40]: I thought this was still an open question; for example, generated entity models, currently created by the RifLayoutsProcessor.java, would require a pretty significant effort to meet this constraint.

**Review by mackec1 — APPROVED** [review]
Just had the one comment on the breadth of what this PR encompassses.

File: docs/style/style-guide.md (hunk 2 above)
Comment (lsmitchell) [reply docs/style/style-guide.md:40]: I mentioned it in the guide as well, but generated classes are excluded from these constraints.

**Pull request merged**